File: pipeline/dp_cpp/src/route.cpp
#include "route.hpp"
#include <algorithm>
#include <cmath>
#include <stdexcept>
#include <yaml-cpp/yaml.h>

// ---- Route methods ----

double Route::length_nm() const {
    if (windows.empty()) return 0.0;
    double s = 0.0;
    for (auto& seg : windows[0].segments) s += seg.distance;
    return s;
}

double Route::eta_h() const {
    return windows.empty() ? 0.0 : windows.back().end;
}

std::vector<double> Route::cumulative_segment_endpoints() const {
    if (windows.empty()) return {};
    std::vector<double> out;
    double cum = 0.0;
    const auto& segs = windows[0].segments;
    for (int i = 0; i + 1 < (int)segs.size(); ++i) {
        cum += segs[i].distance;
        out.push_back(cum);
    }
    return out;
}

const Segment& Route::segment_for_distance(double d, int window_idx) const {
    double cum = 0.0;
    const auto& segs = windows[window_idx].segments;
    for (auto& s : segs) {
        cum += s.distance;
        if (d <= cum + 1e-9) return s;
    }
    return segs.back();
}

const ForecastWindow& Route::window_for_time(double t) const {
    for (auto& w : windows)
        if (w.start - 1e-9 <= t && t < w.end - 1e-9) return w;
    return windows.back();
}

const Segment& Route::weather_at(double t, double d) const {
    const auto& w = window_for_time(t);
    double cum = 0.0;
    for (auto& s : w.segments) {
        cum += s.distance;
        if (d <= cum + 1e-9) return s;
    }
    return w.segments.back();
}

// ---- YAML loader ----

Route load_yaml_route(const std::string& path) {
    YAML::Node data = YAML::LoadFile(path);
    Route route;
    for (auto wnode : data["forecasts"]) {
        ForecastWindow fw;
        auto fwmap = wnode["forecast_window"];
        fw.start = fwmap["start"].as<double>();
        fw.end   = fwmap["end"].as<double>();
        for (auto snode : wnode["segments_table"]) {
            Segment seg;
            seg.id           = snode["id"].as<int>();
            seg.distance     = snode["distance"].as<double>();
            seg.ship_heading = snode["ship_heading"].as<double>();
            seg.wind_dir     = snode["wind_dir"].as<double>();
            seg.beaufort     = snode["beaufort"].as<int>();
            seg.wave_height  = snode["wave_height"].as<double>();
            seg.current_dir  = snode["current_dir"].as<double>();
            seg.current_speed= snode["current_speed"].as<double>();
            fw.segments.push_back(seg);
        }
        route.windows.push_back(fw);
    }
    return route;
}

Route synthesize_multi_window(const Route& route, double window_h) {
    if (route.windows.empty()) return route;
    const auto& base_segs = route.windows[0].segments;
    double eta = route.eta_h();
    Route out;
    double t = 0.0;
    while (t < eta - 1e-9) {
        double t_end = std::min(t + window_h, eta);
        ForecastWindow fw;
        fw.start    = t;
        fw.end      = t_end;
        fw.segments = base_segs;
        out.windows.push_back(fw);
        t = t_end;
    }
    return out;
}

// ---- Waypoint YAML loader ----

std::pair<Route, std::vector<Waypoint>>
build_route_from_waypoints_yaml(const std::string& yaml_path,
                                 std::optional<double> eta_h_opt,
                                 double cruise_sog_kn) {
    YAML::Node data = YAML::LoadFile(yaml_path);
    auto raw = data["waypoints"];
    if (!raw || raw.size() < 2)
        throw std::runtime_error("Route YAML must have >= 2 waypoints");

    std::vector<Waypoint> wps;
    int idx = 1;
    for (auto node : raw) {
        Waypoint wp;
        wp.idx     = idx++;
        wp.lat_deg = node["lat"].as<double>();
        wp.lon_deg = node["lon"].as<double>();
        if (node["name"]) wp.name = node["name"].as<std::string>();
        wps.push_back(wp);
    }

    std::vector<Segment> segs;
    double total_d = 0.0;
    for (int i = 0; i + 1 < (int)wps.size(); ++i) {
        Segment s;
        s.id           = i + 1;
        s.distance     = rhumb_distance_nm(wps[i].lat_deg, wps[i].lon_deg,
                                            wps[i+1].lat_deg, wps[i+1].lon_deg);
        s.ship_heading = rhumb_bearing_deg(wps[i].lat_deg, wps[i].lon_deg,
                                            wps[i+1].lat_deg, wps[i+1].lon_deg);
        s.wind_dir = s.beaufort = 0;
        s.wave_height = s.current_dir = s.current_speed = 0.0;
        total_d += s.distance;
        segs.push_back(s);
    }
    double eta = eta_h_opt.value_or(total_d / cruise_sog_kn);
    ForecastWindow fw{0.0, eta, segs};
    Route route;
    route.windows.push_back(fw);
    return {route, wps};
}

// ---- Hard-coded paper waypoints (route_waypoints.py) ----

const std::vector<Waypoint> WAYPOINTS = {
    {1, 24.75, 52.83, "Port A (Persian Gulf)", 1,  61.25, 223.86, 12.7, 25.54, 139.0, 3, 1.0, 245.0, 0.30},
    {2, 26.55, 56.45, "Gulf of Oman",          2, 121.53, 282.54, 12.6, 31.93, 207.0, 3, 1.0, 248.0, 0.72},
    {3, 24.08, 60.88, std::nullopt,             3, 117.61, 303.18, 12.7, 32.33,   9.0, 4, 1.5, 158.0, 0.73},
    {4, 21.73, 65.73, std::nullopt,             4, 139.03, 298.44, 12.5, 32.18, 201.0, 4, 1.5, 178.0, 0.21},
    {5, 17.96, 69.19, std::nullopt,             5, 143.63, 280.51, 12.3, 31.66,  88.0, 5, 2.5, 135.0, 0.49},
    {6, 14.18, 72.07, std::nullopt,             6, 140.84, 287.34, 12.2, 32.60,  86.0, 4, 1.5, 113.0, 0.22},
    {7, 10.45, 75.16, std::nullopt,             7, 136.42, 284.40, 12.2, 32.00, 353.0, 3, 1.0, 338.0, 0.54},
    {8,  7.00, 78.46, std::nullopt,             8, 110.37, 233.25, 12.2, 30.74,  35.0, 5, 2.5, 290.0, 1.25},
    {9,  5.64, 82.12, std::nullopt,             9, 102.57, 301.80, 12.8, 33.72, 269.0, 4, 1.5, 270.0, 0.28},
    {10, 4.54, 87.04, std::nullopt,            10,  82.83, 315.70, 12.6, 32.32, 174.0, 3, 1.0,  93.0, 0.72},
    {11, 5.20, 92.27, std::nullopt,            11,  84.87, 293.80, 12.7, 34.41,  60.0, 1, 0.1, 185.0, 0.62},
    {12, 5.64, 97.16, std::nullopt,            12, 142.39, 288.42, 12.3, 31.57, 315.0, 3, 1.0,  90.0, 0.30},
    {13, 1.81,100.10, "Port B (Strait of Malacca)", std::nullopt, std::nullopt, std::nullopt,
                       std::nullopt, std::nullopt, std::nullopt, std::nullopt, std::nullopt,
                       std::nullopt, std::nullopt},
};

File: paper_workspace/scripts/dp_rebuild/route.py
"""
DP graph rebuild — route loader.

Parses the legacy `weather_forecasts.yaml` format into a `Route` object.
Dropped fields: `current_angle` (Q_yaml_2: derived duplicate of `current_dir`).
Parser behavior for segment 6 duplicate `distance` (Q_yaml_3): YAML spec says
the last key wins → 287.34.
"""

from dataclasses import dataclass
from pathlib import Path
from typing import Callable, List, Optional

import yaml


@dataclass
class Segment:
    id: int
    distance: float         # nm
    ship_heading: float     # deg
    wind_dir: float         # deg
    beaufort: int
    wave_height: float      # m
    current_dir: float      # deg
    current_speed: float    # knots


@dataclass
class ForecastWindow:
    start: float            # hours
    end: float              # hours
    segments: List[Segment]


@dataclass
class Route:
    windows: List[ForecastWindow]

    @property
    def length_nm(self) -> float:
        if not self.windows:
            return 0.0
        return sum(s.distance for s in self.windows[0].segments)

    @property
    def eta_h(self) -> float:
        if not self.windows:
            return 0.0
        return self.windows[-1].end

    def cumulative_segment_endpoints(self) -> List[float]:
        """Cumulative distance at each segment boundary (excluding the route endpoint).

        For 12 segments this returns 11 interior boundary distances; the terminal
        endpoint at L is added separately by the H-line builder.
        """
        if not self.windows:
            return []
        out: List[float] = []
        cum = 0.0
        segs = self.windows[0].segments
        for s in segs[:-1]:
            cum += s.distance
            out.append(cum)
        return out

    def segment_for_distance(self, d: float, window_idx: int = 0) -> Segment:
        """Return the segment containing distance d along the route."""
        cum = 0.0
        segs = self.windows[window_idx].segments
        for s in segs:
            cum += s.distance
            if d <= cum + 1e-9:
                return s
        return segs[-1]

    def window_for_time(self, t: float) -> ForecastWindow:
        """Return forecast window whose [start, end) contains t. Clamps to last."""
        for w in self.windows:
            if w.start - 1e-9 <= t < w.end - 1e-9:
                return w
        return self.windows[-1]

    def weather_at(self, t: float, d: float) -> Segment:
        """Weather = segment-at-d within the forecast-window-at-t."""
        w = self.window_for_time(t)
        cum = 0.0
        for s in w.segments:
            cum += s.distance
            if d <= cum + 1e-9:
                return s
        return w.segments[-1]


def load_yaml_route(path: Path | str) -> Route:
    """Load a weather_forecasts.yaml file into a Route."""
    with open(path) as f:
        data = yaml.safe_load(f)
    windows: List[ForecastWindow] = []
    for w in data.get("forecasts", []) or []:
        fw = w.get("forecast_window", {}) or {}
        start = float(fw["start"])
        end = float(fw["end"])
        segments: List[Segment] = []
        for s in w.get("segments_table", []) or []:
            segments.append(Segment(
                id=int(s["id"]),
                distance=float(s["distance"]),
                ship_heading=float(s["ship_heading"]),
                wind_dir=float(s["wind_dir"]),
                beaufort=int(s["beaufort"]),
                wave_height=float(s["wave_height"]),
                current_dir=float(s["current_dir"]),
                current_speed=float(s["current_speed"]),
                # current_angle intentionally dropped (Q_yaml_2)
            ))
        windows.append(ForecastWindow(start=start, end=end, segments=segments))
    return Route(windows=windows)


def build_route_from_waypoints_yaml(
    yaml_path: "Path | str",
    eta_h: Optional[float] = None,
    cruise_sog_kn: float = 12.0,
):
    """Build (Route, List[Waypoint]) from a waypoint-only YAML.

    For routes that DON'T ship with paper β / segment metadata (Route 2 onwards).
    Bearings are computed via rhumb_bearing_deg between consecutive waypoints;
    distances via rhumb_distance_nm. Weather fields on each Segment are
    placeholder zeros — the rebuild reads weather from the HDF5 cell-canonical
    lookup, not from the segment.

    YAML schema (`pipeline/config/routes/<name>.yaml`):
        name: <human label>
        description: …
        waypoints:
          - {lat: …, lon: …, name: <optional>}
          - …

    `eta_h`: voyage ETA in hours. If None, defaults to total rhumb / cruise_sog_kn.

    Returns `(route, waypoints)` — `route` is a single-window Route ready for
    `synthesize_multi_window`; `waypoints` is the canonical Waypoint list used
    by `frame.from_route`, `geo_grid.position_at_d`, etc.
    """
    # Local imports keep load_route importable without geo_grid / route_waypoints.
    from geo_grid import rhumb_bearing_deg, rhumb_distance_nm
    from route_waypoints import Waypoint

    with open(yaml_path) as f:
        data = yaml.safe_load(f)

    raw_wps = data.get("waypoints", []) or []
    if len(raw_wps) < 2:
        raise ValueError(f"Route YAML {yaml_path} must list ≥ 2 waypoints, got {len(raw_wps)}")

    waypoints: List[Waypoint] = [
        Waypoint(idx=i + 1,
                 lat_deg=float(wp["lat"]), lon_deg=float(wp["lon"]),
                 name=wp.get("name"))
        for i, wp in enumerate(raw_wps)
    ]

    segments: List[Segment] = []
    total_d = 0.0
    for i in range(len(waypoints) - 1):
        w1, w2 = waypoints[i], waypoints[i + 1]
        d = rhumb_distance_nm(w1.lat_deg, w1.lon_deg, w2.lat_deg, w2.lon_deg)
        b = rhumb_bearing_deg(w1.lat_deg, w1.lon_deg, w2.lat_deg, w2.lon_deg)
        segments.append(Segment(
            id=i + 1,
            distance=d,
            ship_heading=b,
            # Placeholder weather fields — ignored by the rebuild (HDF5 wins).
            wind_dir=0.0, beaufort=0, wave_height=0.0,
            current_dir=0.0, current_speed=0.0,
        ))
        total_d += d

    if eta_h is None:
        eta_h = total_d / cruise_sog_kn

    window = ForecastWindow(start=0.0, end=float(eta_h), segments=segments)
    return Route(windows=[window]), waypoints


def load_route_auto(
    yaml_path: "Path | str",
    eta_h: Optional[float] = None,
    cruise_sog_kn: float = 12.0,
):
    """Dispatcher: pick the right loader from the YAML schema.

    `forecasts:`  → legacy segments-table (paper Persian Gulf) + hardcoded WAYPOINTS.
    `waypoints:`  → lat/lon list (e.g. Atlantic); distances + headings computed.

    Returns `(route, waypoints)` always — for the legacy path the hardcoded
    paper waypoints are returned so callers can stop depending on a global.
    """
    with open(yaml_path) as f:
        data = yaml.safe_load(f)
    if data.get("forecasts"):
        from route_waypoints import WAYPOINTS
        return load_yaml_route(yaml_path), list(WAYPOINTS)
    if data.get("waypoints"):
        return build_route_from_waypoints_yaml(yaml_path, eta_h=eta_h,
                                                cruise_sog_kn=cruise_sog_kn)
    raise ValueError(
        f"Route YAML {yaml_path} must have either 'forecasts:' or 'waypoints:' "
        "top-level key")


def synthesize_multi_window(
    route: Route,
    window_h: float = 6.0,
    noise_fn: Optional[Callable[[Segment, int], Segment]] = None,
) -> Route:
    """Replicate a single-window route into multiple equal windows of `window_h`.

    Q_yaml_4: exercise time-varying-weather infrastructure even though the YAML
    ships with one long window. Default behavior: copy same segments into each
    window (static weather). Supply `noise_fn(segment, window_idx) -> Segment`
    to inject per-window variation.

    Last window is trimmed so its end equals route.eta_h exactly.
    """
    if not route.windows:
        return route
    base_segments = route.windows[0].segments
    eta = route.eta_h

    new_windows: List[ForecastWindow] = []
    t = 0.0
    idx = 0
    while t < eta - 1e-9:
        t_end = min(t + window_h, eta)
        if noise_fn is None:
            segs = list(base_segments)
        else:
            segs = [noise_fn(s, idx) for s in base_segments]
        new_windows.append(ForecastWindow(start=t, end=t_end, segments=segs))
        t = t_end
        idx += 1
    return Route(windows=new_windows)


def summarize_route(route: Route) -> None:
    print("=" * 60)
    print("Route summary")
    print("=" * 60)
    print(f"Length:        {route.length_nm:.2f} nm")
    print(f"ETA:           {route.eta_h:.1f} h")
    print(f"Windows:       {len(route.windows)}")
    print(f"Segments/window: {len(route.windows[0].segments) if route.windows else 0}")
    if route.windows:
        print("\nSegment boundaries (cumulative, nm):")
        cum = 0.0
        for s in route.windows[0].segments:
            cum += s.distance
            print(f"  id {s.id:2d}: +{s.distance:7.2f} -> {cum:7.2f} nm  "
                  f"heading {s.ship_heading:6.2f}°  BN{s.beaufort}")
    print("=" * 60)


if __name__ == "__main__":
    yaml_path = Path(__file__).resolve().parent.parent / \
        "config" / "routes" / "persian_gulf_malacca_paper.yaml"
    route = load_yaml_route(yaml_path)
    summarize_route(route)

    print()
    multi = synthesize_multi_window(route, window_h=6.0)
    print(f"After synthesize_multi_window(6 h): {len(multi.windows)} windows, "
          f"first = [{multi.windows[0].start}, {multi.windows[0].end}), "
          f"last = [{multi.windows[-1].start}, {multi.windows[-1].end})")
